### PR TITLE
Adding Additional Functionality to Neural Quantum States (NQS) for Quantum Chemistry Repo

### PR DESCRIPTION
**Pull request by knitterAQ**
This PR is intended to incorporate two new features into the existing NQS repo in Tangelo-Examples. It is our hope that these features will help make this repo a more complete codebase for general autoregressive NQS testing for quantum chemistry.

1. A new model class is created, implementing a variation of the transformer-based NQS ansatz from a paper by Bennewitz et al. concerning neural quantum state tomography. This variation differs form the NNQS-Transformer version by restricting the phase function to a single linear layer mapping a concatenation of the transformer's hidden dimension outputs for an entire sequence of input tokens to a single phase value.
2. A new objective class, Surrogate, implementing an approximate local energy calculation from the NNQS-Transformer paper by Wu et al. Instead of exactly calculating local energy values, as in the standard Automatic Hamiltonian class, Surrogate approximates local energies at each gradient update step by only considering contributions from terms that correspond with samples generated from the ansatz's state distribution at that step.

Additional smaller changes were made to the main train functions and a few other files to accommodate the form of this new Surrogate class.

### CODE REVIEW

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from .base import Base
+
+class AltTransformer(Base):
+    def __init__(self, num_sites: int, num_spin_up: int, num_spin_down: int, embedding_dim: int=16, nhead: int=2, dim_feedforward: int=64, num_layers: int=1, temperature: float=1.0, device: str=None, **kwargs):
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:9]: I would prefer documentation for all arguments

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            num_layers: number of transformer blocks
+            temperature: modulus network softmax temperature parameter
+            device: device to store model on
+        '''
+        super(AltTransformer, self).__init__('AltTransformer', num_sites, num_spin_up, num_spin_down, device)
+
+        # construct model
+        self.num_in, self.num_out = num_sites, num_sites*2
+        self.temperature = temperature
+        self.input_order = np.stack([np.arange(self.num_sites-2,-1,-2), np.arange(self.num_sites-1,-1,-2)],1).reshape(-1) # [4,5,2,3,0,1]
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:25]: Can it raise problems if `input_order` are not int? Another way of doing this without numpy could be
```python
list(chain.from_iterable(zip(range(n-2, -1, -2), range(n-1, -1, -2))))
```
But I think your solution is OK.

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        self.input_order = torch.Tensor(self.input_order).int().to(self.device)
+        self.shell_order = torch.arange(self.num_sites//2-1, -1, -1) # [2,1,0]
+        
+        transformer_layer = nn.TransformerEncoderLayer(embedding_dim, nhead, dim_feedforward=dim_feedforward, dropout=0.0, batch_first=True)
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:29]: A comment above for this block of code would help identify what's this code block for.

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+        self.net_phase = nn.Linear(in_features=embedding_dim*len(self.shell_order), out_features=4, bias=True)
+
+        self.mask = torch.zeros((len(self.shell_order), len(self.shell_order))).to(self.device)
+        for i in range(len(self.mask)):
+            for j in range(len(self.mask)):
+                if i < j:
+                    self.mask[i][j] = float('-inf') 
+        
+    def _init_weights(self, module: nn.Module):
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:46]: Missing documentation

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        input[:,1:] = shells[:,:-1]
+        pos = self.shell_order.to(self.device)
+
+        input = self.tok_emb(input) + self.pos_emb(pos)
+        # new x is of shape (batch_size, sequence_length, d_model)
+
+        if self.mask.device != self.device:
+            self.mask = self.mask.to(self.device)
+        output = self.transformer(input[:,:(len(self.shell_order) - sample_shell + 1)], mask=self.mask[:(len(self.shell_order) - sample_shell + 1),:(len(self.shell_order) - sample_shell + 1)], is_causal=True)
+        if not self.sampling:
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:78]: I would recommend adding some spaces in-between `if` blocks for better readability.

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            log_psi_cond = 0.5 * self.log_softmax(logits_cls)
+            idx = self.state2shell(x)
+            log_psi_real = log_psi_cond.gather(-1, idx.unsqueeze(-1)).sum(-1).sum(-1)
+            log_psi_imag = self.net_phase(phase_input).gather(-1, idx[:, -1].unsqueeze(-1)).squeeze()
+            if log_psi_real.shape[0] == 1:
+                log_psi_imag = log_psi_imag.reshape(log_psi_real.shape)
+            log_psi = torch.stack((log_psi_real, log_psi_imag), dim=-1)
+            return log_psi
+
+    def apply_constraint(self, inp, log_psi_cond):
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:103]: Missing docs and type hinting

File: examples/neural_quantum_states/src/models/alt_transformer.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        # first entry must be up
+        log_psi_cond[:,:,1].masked_fill_(condition2_up[:,idx]==1, float('-inf'))
+        log_psi_cond[:,:,3].masked_fill_(condition2_up[:,idx]==1, float('-inf'))
+        # second entry must be up
+        log_psi_cond[:,:,2].masked_fill_(condition2_down[:,idx]==1, float('-inf'))
+        log_psi_cond[:,:,3].masked_fill_(condition2_down[:,idx]==1, float('-inf'))
+        return log_psi_cond
+
+    @torch.no_grad()
+    def sample(self, bs, num_samples):
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/models/alt_transformer.py:135]: Missing doc and type hinting

File: examples/neural_quantum_states/src/objective/surrogate.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+from tqdm import tqdm
+from src.complex import exp, scalar_mult
+from .hamiltonian import Hamiltonian
+
+
+class Surrogate(Hamiltonian):
+    def __init__(self, hamiltonian_string: str, num_sites: int, flip_bs: int, **kwargs):
+        '''
+        An implemenation of the surrogate Hamiltonian described in https://dl.acm.org/doi/10.1145/3581784.3607061
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/objective/surrogate.py:14]: typo

File: examples/neural_quantum_states/src/objective/surrogate.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        del self.operators
+        # track the exponential of -i
+        self.num_pauli_y = pauli_y_idx.sum(-1) # [num_terms]
+        part1 = (-1j)**self.num_pauli_y.detach()
+        part1 = torch.stack((part1.real, part1.imag), dim=-1).float()
+        self.coefficients = scalar_mult(self.coefficients, part1)
+        del part1
+        # the unique element has flipped value if the corresponding pauli is x or y.
+        flip_idx = pauli_x_idx + pauli_y_idx # [num_terms, input_dim]
+        # self.flip_idx = flip_idx
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/objective/surrogate.py:38]: Commented code here, and spaces should be added for better readability. Also, why do you need to `del` some variables?

File: examples/neural_quantum_states/src/objective/surrogate.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        ratio = exp(log_psi_k-log_psi.unsqueeze(1)).detach() # [bs, num_terms, 2]
+        # compute matrix element
+        # Eq. B3
+        part2 = (x.unsqueeze(1).repeat(1, self.num_terms, 1) * self.select_idx.unsqueeze(0) + (1-self.select_idx).unsqueeze(0)).prod(-1) # [bs, num_terms]
+        mtx_k = torch.stack((part2, torch.zeros_like(part2)), dim=-1)
+        # total local energy
+        local_energy = scalar_mult(self.coefficients.unsqueeze(0), scalar_mult(mtx_k, ratio)).sum(1) # [bs, 2]
+        return local_energy.detach(), log_psi
+                            
+    def set_device(self, device: str):
```
Comment (alexfleury-sb) [inline examples/neural_quantum_states/src/objective/surrogate.py:189]: Missing argument in docs

**Review by alexfleury-sb — CHANGES_REQUESTED** [review]
Great work! I can't say much about the algorithm itself, but I added few comments to improve readability.

File: examples/neural_quantum_states/src/models/alt_transformer.py (hunk 2 above)
Comment (knitterAQ) [reply examples/neural_quantum_states/src/models/alt_transformer.py:25]: This NQS codebase is only intended for second quantized molecular Hamiltonians, so num_sites will always be an even number and there shouldn't be any problem.

File: examples/neural_quantum_states/src/objective/surrogate.py (hunk 9 above)
Comment (knitterAQ) [reply examples/neural_quantum_states/src/objective/surrogate.py:41]: Intermediate data structures for the Hamiltonian are explicitly deleted to preserve as much available memory as possible.
